### PR TITLE
DEP: deprecate empty field names

### PR DESCRIPTION
**Pull request by mattip**
Split off from PR #12358. Empty field names are confusing, users should construct record arrays with name, format, offset dictionaries instead.

### CODE REVIEW

**Comment by charris** [comment]
Labels please.

File: numpy/lib/tests/test_stride_tricks.py [hunk 1]
```diff
@@ -324,7 +325,9 @@ def test_as_strided():
     assert_equal(a.dtype, a_view.dtype)
 
     # Make sure that the only type that could fail is properly handled
-    dt = np.dtype({'names': [''], 'formats': ['V4']})
+    with warnings.catch_warnings(record=True) as w:
```
Comment (charris) [inline numpy/lib/tests/test_stride_tricks.py]: Do you need the `w`?

Comment (charris) [reply numpy/lib/tests/test_stride_tricks.py]: And if not, why the always? AFAICT, you just want to suppress the warning. For that see `suppress_warnings`. There is also `pytest.mark.filterwarnings`, but I don't know if we trust it.

Comment (mattip) [reply numpy/lib/tests/test_stride_tricks.py]: fixed to use suppress_warnings

**Comment by mattip** [comment]
should hit the mailing list

**Comment by charris** [comment]
If it needs to hit the mailing list, please do so.

**Comment by mattip** [comment]
Hmm. The problem is deeper than just blank names. `dtype.descr`does not properly handle overlapping fields:
``` python
>>>a = np.dtype({'names': ['a', 'b'], 'formats': ['i4'] * 2, 'offsets': [0, 2]})
>>>a.descr
[('a', '<i4'), ('b', '<i4')]  # loses the offsets :(
>>>str(a)
"{'names':['a','b'], 'formats':['<i4','<i4'], 'offsets':[0,2], 'itemsize':6}"
```
Maybe we should abandon deprecating blank fields, as it does not solve the roundtrip save/load problem. We could try using `str(dtype)` in saving/loading record arrays, as it seems to capture the full expression of the dtype.

**Comment by eric-wieser** [comment]
> dtype.descrdoes not properly handle overlapping fields:

On 1.15 your code gives an error,
```
ValueError: dtype.descr is not defined for types with overlapping or out-of-order fields
```

**Comment by eric-wieser** [comment]
> as it seems to capture the full expression of the dtype.

Can't we just pickle the dtype?

**Comment by ahaldane** [comment]
Also, that's similar to the overlapping-fields problem in the PEP3118 interface, see [this comment](https://github.com/numpy/numpy/pull/7798#discussion_r123321622). 

It turns out that the PEP3118 structure format string cannot represent overlapping or out-of-order fields. Since `dtype.descr` was originally designed for use in `__array_interface__`, which is essentially a forerunner to the PEP3118-interface, it isn't surprising it can't represent overlapping fields either.

In my opinion, the "best" ultimate behavior would be for us to work with CPython to update or extend the PEP3118 format string for structs so it can handle overlapping fields, out-of-order-fields, and padding, using the justification in the linked comment above. Then deprecate `__array_function__` and use PEP3118 everywhere, including in `np.save/load` when serializing arrays. Currently `.descr` is used to serialize the dtype - it would be nice to use PEP3118 datatype notation instead since it was in principle designed for such serialization.

**Comment by eric-wieser** [comment]
`s/__array_function__/__array_interface__/`?

**Comment by eric-wieser** [comment]
> and use PEP3118 everywhere, including in np.save/load when serializing arrays.

It's not clear to me why the buffer interface and a serialization format need to be related. It's not unreasonable for numpy to support custom or third-party dtypes that are not part of PEP3118, and we need to be able to save those too.

Python already has a serialization protocol - pickle. Why can't we just use it? What makes it worth defining a custom serialization format for the data type?

**Comment by ahaldane** [comment]
Well, I suppose that's true, that seems like a reasonable solution as well. We've already written all the pickle serialization stuff in `descriptor.c`.

My background thought, though, was it would be "cool" if `.npy` files were simply gzipped PEP3118 buffers, which anyone (and not just numpy) could read.

**Comment by charris** [comment]
@mattip Abandon?

**Pull request closed without merging**